pkg/policy/directory/watcher.go

```diff
@@ -0,0 +1,207 @@
+	if err != nil {
+		return err
+	}
+
+	//update labels
+	for _, r := range rules {
+		r.Labels = getLabels(cnpFilePath, cnp)
+	}
+
+	// add to policy engine
+	_, err = p.policyManager.PolicyAdd(rules, &policy.AddOptions{
+		ReplaceWithLabels: getLabels(cnpFilePath, cnp),
+		Source:            source.Directory,
+		Resource:          resourceID,
+	})
+
+	if err == nil {
+		p.fileNameToCnpCache[cnpFilePath] = cnp
+	}
+
+	return err
+}
+
+func (p *policyWatcher) watchDirectory(ctx context.Context) {
+	go func() {
+		watcher, err := fsnotify.NewWatcher()
+		if err != nil {
+			p.log.WithField("Err", err).Error("Initializing NewWatcher failed")
+			return
+		}
+		defer watcher.Close()
+
+		dir := p.config.StaticCiliumNetworkPolicyPath
+		err = watcher.Add(dir)
+		if err != nil {
+			p.log.WithFields(logrus.Fields{"Dir": dir, "err": err}).Error("Directory Watcher Add failed")
+			return
+		}
+
+		files, err := os.ReadDir(dir)
+		if err != nil {
+			p.log.WithFields(logrus.Fields{"Directory": dir, "err": err}).Warn("Error reading CNP directory")
+			return
+		}
+
+		// Read existing cnp files before watcher was added
+		for _, f := range files {
+			err := p.addToPolicyEngine(filepath.Join(dir, f.Name()))
+			if err != nil {
+				p.log.WithFields(logrus.Fields{"file": f.Name(), "err": err}).Error("Error while adding network policy to policy engine")
+			} else {
+
+			}
+			reportCNPChangeMetrics(err)
+		}
+
+		// Listen for file add, update and delete
+		for {
+			select {
+			case event := <-watcher.Events:
+				if !(event.Op&fsnotify.Create == fsnotify.Create ||
+					event.Op&fsnotify.Remove == fsnotify.Remove) {
+					continue
+				}
+				if filepath.Ext(event.Name) != ".yaml" {
+					continue
+				}
+				if event.Op&fsnotify.Create == fsnotify.Create {
+					p.log.WithField("file", event.Name).Debug("CNP file added/updated in directory")
+					err := p.addToPolicyEngine(event.Name)
+					if err != nil {
+						p.log.WithFields(logrus.Fields{"file": event.Name, "err": err}).Error("Error while adding network policy to policy engine")
+					}
+				}
+				if event.Op&fsnotify.Remove == fsnotify.Remove {
+					p.log.WithField("file", event.Name).Debug("CNP file removed from directory")
+					cnp := p.fileNameToCnpCache[event.Name]
+					if cnp == nil {
+						p.log.WithField("file", event.Name).Debug("map entry doesn't exist")
+						continue
+					}
+
+					resourceID := ipcacheTypes.NewResourceID(
+						ipcacheTypes.ResourceKindCNP,
+						cnp.ObjectMeta.Namespace,
+						cnp.ObjectMeta.Name,
+					)
+					_, err = p.policyManager.PolicyDelete(getLabels(event.Name, cnp), &policy.DeleteOptions{
+						Source:   source.Directory,
+						Resource: resourceID})
+					if err == nil {
+						delete(p.fileNameToCnpCache, event.Name)
+					} else {
+						p.log.WithField("err", err).Error("Deleting network policy failed")
+					}
+				}
+				reportCNPChangeMetrics(err)
+			case err := <-watcher.Errors:
+				p.log.Error("Error:", err)
```

Comment:
can you please clarify how should we handle ErrEventOverflow? I saw increasing buffer size is one option to auto recover but suggested method works only for windows.
`This only has effect on Windows systems, and is a no-op for other backends.`

I would rather keep simple and return error instead of increasing buffer size which may have other complications

Reply:
Ah ok, I see what you mean. Can we just change the `Error:` string to something like `Unexpected error thrown by fsnotify watcher when watching policy directory` then?

Reply:
I'd generally expect the form `p.log.WithError(err).Error(...)` to ensure proper structured logging, too.

Reply:
yep agreed. I addressed that